Sources/SwiftFusion/Core/FixedShapeTensor.swift

```diff
@@ -0,0 +1,77 @@
+import TensorFlow
+
+public protocol FixedShapeTensor: EuclideanVectorN {
+  static var shape: TensorShape { get }
+  @differentiable init(_ tensor: Tensor<Double>)
+  @differentiable var tensor: Tensor<Double> { get set }
+}
+
+/// Default implementations of `EuclideanVectorN` requirements.
+extension FixedShapeTensor {
+  @differentiable
+  public static func += (_ lhs: inout Self, _ rhs: Self) {
+    lhs = Self(lhs.tensor + rhs.tensor)
+  }
+
+  @differentiable
+  public static func -= (_ lhs: inout Self, _ rhs: Self) {
+    lhs = Self(lhs.tensor - rhs.tensor)
+  }
+
+  @differentiable
+  public static func *= (_ lhs: inout Self, _ rhs: Double) {
+    lhs = Self(lhs.tensor * rhs)
+  }
+
+  @differentiable
+  public func dot(_ other: Self) -> Double {
+    (self.tensor * other.tensor).sum().scalarized()
+  }
+
+  public init<Source: Collection>(_ scalars: Source) where Source.Element == Double {
+    self.init(Tensor(shape: Self.shape, scalars: Array(scalars)))
+  }
+
+  public static var dimension: Int {
+    shape.reduce(1, *)
+  }
+
+  public static var standardBasis: [Self] {
+    (0..<dimension).map { i in
+      var b = Array(repeating: Double(0), count: dimension)
+      b[i] = 1
+      return Self(Tensor(shape: shape, scalars: b))
+    }
+  }
+
+  /// Returns the result of calling `body` on the scalars of `self`.
+  public func withUnsafeBufferPointer<R>(
+    _ body: (UnsafeBufferPointer<Double>) throws -> R
+  ) rethrows -> R {
+    try self.tensor.scalars.withUnsafeBufferPointer(body)
+  }
+
+  /// Returns the result of calling `body` on the scalars of `self`.
+  public mutating func withUnsafeMutableBufferPointer<R>(
+    _ body: (UnsafeMutableBufferPointer<Double>) throws -> R
+  ) rethrows -> R {
+    var scalars = self.tensor.scalars
+    let r = try scalars.withUnsafeMutableBufferPointer { b in
+      try body(b)
+    }
+    self.tensor = Tensor(shape: Self.shape, scalars: scalars)
+    return r
+  }
+}
+
+public struct Tensor10_10: AdditiveArithmetic, FixedShapeTensor {
```

Comment:
a bit weird to have just 10,10 here. Is that for a test?

Reply:
It's for my PPCA, this is still very experimental, so I'll merge now and improve later.

Reply:
currently `EuclideanVectorN` (which is used in all the factor graph infra) has a requirement that the size is known statically at compiletime. the easiest way to do this is to define separate types for all the shapes that we want to use. I have some simplifications in mind that might let us lift the restriction that the size be known at compiletime, so I'll try to remove this weird type by implementing those.